segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java

```diff
@@ -185,14 +185,53 @@ public String toString() {
 
     //region ReadLedger
 
-    @RequiredArgsConstructor
+    private final static class EntryHolder {
+        final long entryId;
+        final byte[] entryContent;
+
+        private EntryHolder(LedgerEntry entry) {
+            this.entryId = entry.getEntryId();
+            // we must perform a copy
+            // a LedgerEntry is a wrapper over a refcounted Netty ByteBuf
+            // we do not have control over the lifecycle
+            // of the InputStream returned by the ReadItem interface,
+            // so we perform a copy and then release memory back to Netty
+            this.entryContent = entry.getEntryBytes();
+        }
+    }
+
     private static class ReadLedger {
         final LedgerMetadata metadata;
-        final LedgerHandle handle;
-        final Enumeration<LedgerEntry> reader;
+        final ReadHandle handle;
+        /**
+         * Holder for the list of entries.
+         * A null value means that this ledger is empty.
+         */
+        final Iterator<EntryHolder> entries;
+
+        public ReadLedger(LedgerMetadata metadata, ReadHandle handle, LedgerEntries ledgerEntries) {
+            this.metadata = metadata;
+            this.handle = handle;
+            if (ledgerEntries != null) {
+                List<EntryHolder> entriesHolder = new ArrayList<>();
```

Comment:
Why do you buffer all the entries from the ledger? This, in combination with copying all the data to Java byte arrays and a max ledger size of 1GB has the potential for HeapOOMs.